File: services/modelRuntime/officialConversation.ts
import type { AssistantToolRequest } from './canonical.js'

export type OfficialAssistantToolRequest = AssistantToolRequest & {
  input: Record<string, unknown>
  type: 'tool_use'
}

export type OfficialToolResultBlock = {
  content: unknown
  is_error?: boolean
  tool_use_id: string
  type: 'tool_result'
}

type MessageWithContentLike = {
  message: {
    content: unknown
  }
}

export type AssistantToolRequestWithMessage<
  TMessage extends MessageWithContentLike = MessageWithContentLike,
> = {
  message: TMessage
  toolRequest: OfficialAssistantToolRequest
}

export type OfficialAssistantToolRound = {
  assistantVisibleText: string | undefined
  toolRequests: OfficialAssistantToolRequest[]
  toolResultsByToolUseId: Map<string, OfficialToolResultBlock>
}

export function extractAssistantToolRequests(
  content: unknown,
): OfficialAssistantToolRequest[] {
  if (!Array.isArray(content)) {
    return []
  }

  return content.flatMap(block => {
    if (
      !block ||
      typeof block !== 'object' ||
      !('type' in block) ||
      (block as { type?: string }).type !== 'tool_use'
    ) {
      return []
    }

    const typedBlock = block as {
      id?: unknown
      input?: unknown
      name?: unknown
    }
    if (
      typeof typedBlock.id !== 'string' ||
      typeof typedBlock.name !== 'string' ||
      !typedBlock.input ||
      typeof typedBlock.input !== 'object' ||
      Array.isArray(typedBlock.input)
    ) {
      return []
    }

    return [
      {
        argumentsObject: typedBlock.input as Record<string, unknown>,
        id: typedBlock.id,
        input: typedBlock.input as Record<string, unknown>,
        name: typedBlock.name,
        type: 'tool_use',
      },
    ]
  })
}

export function countAssistantToolRequests(content: unknown): number {
  return extractAssistantToolRequests(content).length
}

export function hasAssistantToolRequests(content: unknown): boolean {
  return countAssistantToolRequests(content) > 0
}

export function findAssistantToolRequestByName(
  content: unknown,
  toolName: string,
): OfficialAssistantToolRequest | undefined {
  return extractAssistantToolRequests(content).find(
    toolRequest => toolRequest.name === toolName,
  )
}

export function hasAssistantToolRequestNamed(
  content: unknown,
  toolName: string,
): boolean {
  return findAssistantToolRequestByName(content, toolName) !== undefined
}

export function extractAssistantToolRequestIds(content: unknown): string[] {
  return extractAssistantToolRequests(content).map(toolRequest => toolRequest.id)
}

export function countAssistantToolRequestsInMessages(
  messages: ReadonlyArray<{ message: { content: unknown } }>,
): number {
  return messages.reduce(
    (total, message) => total + countAssistantToolRequests(message.message.content),
    0,
  )
}

export function extractAssistantToolRequestsInMessages(
  messages: ReadonlyArray<MessageWithContentLike | null | undefined>,
): OfficialAssistantToolRequest[] {
  return messages.flatMap(message =>
    message ? extractAssistantToolRequests(message.message.content) : [],
  )
}

export function extractAssistantToolRequestsWithMessages<
  TMessage extends MessageWithContentLike,
>(
  messages: ReadonlyArray<TMessage | null | undefined>,
): AssistantToolRequestWithMessage<TMessage>[] {
  return messages.flatMap(message =>
    message
      ? extractAssistantToolRequests(message.message.content).map(toolRequest => ({
          message,
          toolRequest,
        }))
      : [],
  )
}

export function findMostRecentAssistantToolRequest(
  messages: ReadonlyArray<MessageWithContentLike | null | undefined>,
  toolName: string,
): OfficialAssistantToolRequest | undefined {
  for (let i = messages.length - 1; i >= 0; i--) {
    const message = messages[i]
    if (!message) {
      continue
    }
    const toolRequest = findAssistantToolRequestByName(
      message.message.content,
      toolName,
    )
    if (toolRequest) {
      return toolRequest
    }
  }
  return undefined
}

export function hasAssistantToolRequestNamedInMessages(
  messages: ReadonlyArray<MessageWithContentLike | null | undefined>,
  toolName: string,
): boolean {
  return findMostRecentAssistantToolRequest(messages, toolName) !== undefined
}

export function extractOfficialToolResultIds(content: unknown): string[] {
  if (!Array.isArray(content)) {
    return []
  }

  return content.flatMap(block =>
    isOfficialToolResultBlock(block) ? [block.tool_use_id] : [],
  )
}

export function extractOfficialToolResultBlocks(
  content: unknown,
): OfficialToolResultBlock[] {
  if (!Array.isArray(content)) {
    return []
  }

  return content.flatMap(block => (isOfficialToolResultBlock(block) ? [block] : []))
}

export function hasOfficialToolResultBlocks(content: unknown): boolean {
  return extractOfficialToolResultBlocks(content).length > 0
}

export function stripOfficialToolResultBlocks(content: unknown): unknown {
  if (!Array.isArray(content)) {
    return content
  }

  const stripped = content.filter(block => !isOfficialToolResultBlock(block))
  return stripped.length === content.length ? content : stripped
}

export function findOfficialToolResultInMessages(
  messages: ReadonlyArray<{ message: { content: unknown }; type: string }>,
  toolUseId: string,
): OfficialToolResultBlock | undefined {
  for (const message of messages) {
    if (message.type !== 'user') {
      continue
    }
    const block = findOfficialToolResultBlock(message.message.content, toolUseId)
    if (block) {
      return block
    }
  }
  return undefined
}

export function extractAssistantVisibleText(content: unknown): string {
  if (!Array.isArray(content)) {
    return ''
  }

  return content
    .flatMap(block =>
      block &&
      typeof block === 'object' &&
      'type' in block &&
      (block as { type?: string }).type === 'text' &&
      'text' in block &&
      typeof (block as { text?: unknown }).text === 'string'
        ? [(block as { text: string }).text]
        : [],
    )
    .join('')
}

export function findMostRecentAssistantVisibleTextInMessages(
  messages: ReadonlyArray<MessageWithContentLike | null | undefined>,
): string | undefined {
  for (let i = messages.length - 1; i >= 0; i--) {
    const message = messages[i]
    if (!message) {
      continue
    }
    const text = extractAssistantVisibleText(message.message.content)
    if (text) {
      return text
    }
  }
  return undefined
}

export function buildOfficialAssistantToolRound(
  assistantMessages: ReadonlyArray<MessageWithContentLike | null | undefined>,
  toolResultMessages: ReadonlyArray<{ message: { content: unknown }; type: string }>,
): OfficialAssistantToolRound {
  const toolRequests = extractAssistantToolRequestsInMessages(assistantMessages)
  const toolResultsByToolUseId = new Map<string, OfficialToolResultBlock>()
  for (const toolRequest of toolRequests) {
    const toolResult = findOfficialToolResultInMessages(
      toolResultMessages,
      toolRequest.id,
    )
    if (toolResult) {
      toolResultsByToolUseId.set(toolRequest.id, toolResult)
    }
  }
  return {
    assistantVisibleText:
      findMostRecentAssistantVisibleTextInMessages(assistantMessages),
    toolRequests,
    toolResultsByToolUseId,
  }
}

export function mapAssistantToolRequestInputs(
  content: unknown,
  mapper: (
    toolRequest: OfficialAssistantToolRequest,
  ) => Record<string, unknown>,
): unknown {
  if (!Array.isArray(content)) {
    return content
  }

  let clonedContent: unknown[] | undefined
  for (let i = 0; i < content.length; i++) {
    const block = content[i]
    const toolRequest = extractAssistantToolRequests([block])[0]
    if (!toolRequest) {
      continue
    }

    const mappedInput = mapper(toolRequest)
    if (mappedInput === toolRequest.input) {
      continue
    }

    clonedContent ??= [...content]
    clonedContent[i] = {
      ...(block as Record<string, unknown>),
      input: mappedInput,
    }
  }

  return clonedContent ?? content
}

export function createOfficialToolResultBlock(
  toolUseId: string,
  content: unknown,
  isError = false,
): OfficialToolResultBlock {
  return {
    content,
    ...(isError ? { is_error: true } : {}),
    tool_use_id: toolUseId,
    type: 'tool_result',
  }
}

export function isOfficialToolResultBlock(
  block: unknown,
): block is OfficialToolResultBlock {
  return (
    block !== null &&
    typeof block === 'object' &&
    'type' in block &&
    (block as { type?: string }).type === 'tool_result' &&
    'tool_use_id' in block &&
    typeof (block as { tool_use_id?: unknown }).tool_use_id === 'string'
  )
}

export function findOfficialToolResultBlock(
  content: unknown,
  toolUseId: string,
): OfficialToolResultBlock | undefined {
  if (!Array.isArray(content)) {
    return undefined
  }

  return content.find(
    block =>
      isOfficialToolResultBlock(block) && block.tool_use_id === toolUseId,
  )
}

export function findFirstOfficialToolResultBlock(
  content: unknown,
): OfficialToolResultBlock | undefined {
  return extractOfficialToolResultBlocks(content)[0]
}

File: crates/mossen-agent/src/services/model_runtime/official_conversation.rs
//! Official conversation — tool request/result extraction and message content utilities.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

use super::canonical::AssistantToolRequest;

/// Official assistant tool request (with input and type).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OfficialAssistantToolRequest {
    pub arguments_object: HashMap<String, Value>,
    pub id: String,
    pub input: HashMap<String, Value>,
    pub name: String,
    #[serde(rename = "type")]
    pub block_type: String,
}

/// Official tool result block.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OfficialToolResultBlock {
    pub content: Value,
    pub is_error: Option<bool>,
    pub tool_use_id: String,
    #[serde(rename = "type")]
    pub block_type: String,
}

/// An assistant tool round with text, requests, and results.
#[derive(Debug, Clone)]
pub struct OfficialAssistantToolRound {
    pub assistant_visible_text: Option<String>,
    pub tool_requests: Vec<OfficialAssistantToolRequest>,
    pub tool_results_by_tool_use_id: HashMap<String, OfficialToolResultBlock>,
}

/// Extract assistant tool requests from content blocks.
pub fn extract_assistant_tool_requests(content: &Value) -> Vec<OfficialAssistantToolRequest> {
    let arr = match content.as_array() {
        Some(a) => a,
        None => return Vec::new(),
    };

    let mut results = Vec::new();
    for block in arr {
        let obj = match block.as_object() {
            Some(o) => o,
            None => continue,
        };
        if obj.get("type").and_then(|v| v.as_str()) != Some("tool_use") {
            continue;
        }
        let id = match obj.get("id").and_then(|v| v.as_str()) {
            Some(s) => s.to_string(),
            None => continue,
        };
        let name = match obj.get("name").and_then(|v| v.as_str()) {
            Some(s) => s.to_string(),
            None => continue,
        };
        let input = match obj.get("input") {
            Some(v) if v.is_object() => {
                serde_json::from_value::<HashMap<String, Value>>(v.clone()).unwrap_or_default()
            }
            _ => continue,
        };

        results.push(OfficialAssistantToolRequest {
            arguments_object: input.clone(),
            id,
            input,
            name,
            block_type: "tool_use".to_string(),
        });
    }
    results
}

/// Count assistant tool requests in content.
pub fn count_assistant_tool_requests(content: &Value) -> usize {
    extract_assistant_tool_requests(content).len()
}

/// Check if content has any tool requests.
pub fn has_assistant_tool_requests(content: &Value) -> bool {
    count_assistant_tool_requests(content) > 0
}

/// Find a tool request by name.
pub fn find_assistant_tool_request_by_name(
    content: &Value,
    tool_name: &str,
) -> Option<OfficialAssistantToolRequest> {
    extract_assistant_tool_requests(content)
        .into_iter()
        .find(|r| r.name == tool_name)
}

/// Check if content has a tool request with a given name.
pub fn has_assistant_tool_request_named(content: &Value, tool_name: &str) -> bool {
    find_assistant_tool_request_by_name(content, tool_name).is_some()
}

/// Extract tool request IDs from content.
pub fn extract_assistant_tool_request_ids(content: &Value) -> Vec<String> {
    extract_assistant_tool_requests(content)
        .into_iter()
        .map(|r| r.id)
        .collect()
}

/// Count tool requests across multiple messages.
pub fn count_assistant_tool_requests_in_messages(messages: &[Value]) -> usize {
    messages.iter().fold(0, |total, msg| {
        let content = msg.pointer("/message/content").unwrap_or(&Value::Null);
        total + count_assistant_tool_requests(content)
    })
}

/// Extract tool requests from multiple messages.
pub fn extract_assistant_tool_requests_in_messages(
    messages: &[Value],
) -> Vec<OfficialAssistantToolRequest> {
    messages
        .iter()
        .flat_map(|msg| {
            let content = msg.pointer("/message/content").unwrap_or(&Value::Null);
            extract_assistant_tool_requests(content)
        })
        .collect()
}

/// Find the most recent tool request by name across messages.
pub fn find_most_recent_assistant_tool_request(
    messages: &[Value],
    tool_name: &str,
) -> Option<OfficialAssistantToolRequest> {
    for msg in messages.iter().rev() {
        let content = msg.pointer("/message/content").unwrap_or(&Value::Null);
        if let Some(req) = find_assistant_tool_request_by_name(content, tool_name) {
            return Some(req);
        }
    }
    None
}

/// Check if any message has a tool request with the given name.
pub fn has_assistant_tool_request_named_in_messages(messages: &[Value], tool_name: &str) -> bool {
    find_most_recent_assistant_tool_request(messages, tool_name).is_some()
}

/// Extract official tool result IDs from content.
pub fn extract_official_tool_result_ids(content: &Value) -> Vec<String> {
    extract_official_tool_result_blocks(content)
        .into_iter()
        .map(|b| b.tool_use_id)
        .collect()
}

/// Extract official tool result blocks from content.
pub fn extract_official_tool_result_blocks(content: &Value) -> Vec<OfficialToolResultBlock> {
    let arr = match content.as_array() {
        Some(a) => a,
        None => return Vec::new(),
    };

    arr.iter()
        .filter_map(|block| {
            if is_official_tool_result_block(block) {
                Some(OfficialToolResultBlock {
                    content: block.get("content").cloned().unwrap_or(Value::Null),
                    is_error: block.get("is_error").and_then(|v| v.as_bool()),
                    tool_use_id: block
                        .get("tool_use_id")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string(),
                    block_type: "tool_result".to_string(),
                })
            } else {
                None
            }
        })
        .collect()
}

/// Check if content has any tool result blocks.
pub fn has_official_tool_result_blocks(content: &Value) -> bool {
    !extract_official_tool_result_blocks(content).is_empty()
}

/// Strip tool result blocks from content.
pub fn strip_official_tool_result_blocks(content: &Value) -> Value {
    let arr = match content.as_array() {
        Some(a) => a,
        None => return content.clone(),
    };

    let stripped: Vec<&Value> = arr
        .iter()
        .filter(|block| !is_official_tool_result_block(block))
        .collect();

    if stripped.len() == arr.len() {
        content.clone()
    } else {
        Value::Array(stripped.into_iter().cloned().collect())
    }
}

/// Find a tool result in messages by tool_use_id.
pub fn find_official_tool_result_in_messages(
    messages: &[Value],
    tool_use_id: &str,
) -> Option<OfficialToolResultBlock> {
    for msg in messages {
        if msg.get("type").and_then(|v| v.as_str()) != Some("user") {
            continue;
        }
        let content = msg.pointer("/message/content").unwrap_or(&Value::Null);
        if let Some(block) = find_official_tool_result_block(content, tool_use_id) {
            return Some(block);
        }
    }
    None
}

/// Extract visible text from assistant content blocks.
pub fn extract_assistant_visible_text(content: &Value) -> String {
    let arr = match content.as_array() {
        Some(a) => a,
        None => return String::new(),
    };

    arr.iter()
        .filter_map(|block| {
            let obj = block.as_object()?;
            if obj.get("type")?.as_str()? == "text" {
                obj.get("text")?.as_str().map(|s| s.to_string())
            } else {
                None
            }
        })
        .collect::<Vec<_>>()
        .join("")
}

/// Find the most recent assistant visible text across messages.
pub fn find_most_recent_assistant_visible_text_in_messages(messages: &[Value]) -> Option<String> {
    for msg in messages.iter().rev() {
        let content = msg.pointer("/message/content").unwrap_or(&Value::Null);
        let text = extract_assistant_visible_text(content);
        if !text.is_empty() {
            return Some(text);
        }
    }
    None
}

/// Build an official assistant tool round from messages.
pub fn build_official_assistant_tool_round(
    assistant_messages: &[Value],
    tool_result_messages: &[Value],
) -> OfficialAssistantToolRound {
    let tool_requests = extract_assistant_tool_requests_in_messages(assistant_messages);
    let mut tool_results_by_tool_use_id: HashMap<String, OfficialToolResultBlock> = HashMap::new();

    for req in &tool_requests {
        if let Some(result) = find_official_tool_result_in_messages(tool_result_messages, &req.id) {
            tool_results_by_tool_use_id.insert(req.id.clone(), result);
        }
    }

    OfficialAssistantToolRound {
        assistant_visible_text: find_most_recent_assistant_visible_text_in_messages(
            assistant_messages,
        ),
        tool_requests,
        tool_results_by_tool_use_id,
    }
}

/// Create an official tool result block.
pub fn create_official_tool_result_block(
    tool_use_id: &str,
    content: Value,
    is_error: bool,
) -> OfficialToolResultBlock {
    OfficialToolResultBlock {
        content,
        is_error: if is_error { Some(true) } else { None },
        tool_use_id: tool_use_id.to_string(),
        block_type: "tool_result".to_string(),
    }
}

/// Check if a block is a tool result block.
pub fn is_official_tool_result_block(block: &Value) -> bool {
    let obj = match block.as_object() {
        Some(o) => o,
        None => return false,
    };
    obj.get("type").and_then(|v| v.as_str()) == Some("tool_result")
        && obj.get("tool_use_id").and_then(|v| v.as_str()).is_some()
}

/// Find a specific tool result block by tool_use_id.
pub fn find_official_tool_result_block(
    content: &Value,
    tool_use_id: &str,
) -> Option<OfficialToolResultBlock> {
    extract_official_tool_result_blocks(content)
        .into_iter()
        .find(|b| b.tool_use_id == tool_use_id)
}

/// Find the first tool result block in content.
pub fn find_first_official_tool_result_block(content: &Value) -> Option<OfficialToolResultBlock> {
    extract_official_tool_result_blocks(content)
        .into_iter()
        .next()
}

/// Map tool request inputs using a mapper function.
pub fn map_assistant_tool_request_inputs<F>(content: &Value, mapper: F) -> Value
where
    F: Fn(&OfficialAssistantToolRequest) -> HashMap<String, Value>,
{
    let arr = match content.as_array() {
        Some(a) => a,
        None => return content.clone(),
    };

    let mut modified = false;
    let mut result: Vec<Value> = arr.clone();

    for (i, block) in arr.iter().enumerate() {
        let requests = extract_assistant_tool_requests(&Value::Array(vec![block.clone()]));
        if let Some(req) = requests.first() {
            let mapped_input = mapper(req);
            if mapped_input != req.input {
                let mut obj = block.as_object().unwrap().clone();
                obj.insert(
                    "input".to_string(),
                    serde_json::to_value(&mapped_input)
                        .unwrap_or(Value::Object(Default::default())),
                );
                result[i] = Value::Object(obj);
                modified = true;
            }
        }
    }

    if modified {
        Value::Array(result)
    } else {
        content.clone()
    }
}

/// A tool request paired with its source message. Mirrors TS
/// `type AssistantToolRequestWithMessage<TMessage> = { message, toolRequest }`.
#[derive(Debug, Clone)]
pub struct AssistantToolRequestWithMessage {
    pub message: Value,
    pub tool_request: OfficialAssistantToolRequest,
}

/// Flatten messages into `(message, tool_request)` pairs. Mirrors TS
/// `extractAssistantToolRequestsWithMessages`.
pub fn extract_assistant_tool_requests_with_messages(
    messages: &[Value],
) -> Vec<AssistantToolRequestWithMessage> {
    let mut out = Vec::new();
    for message in messages {
        // Skip null/undefined entries (rendered as JSON null in the Value form).
        if message.is_null() {
            continue;
        }
        let content = message
            .get("message")
            .and_then(|m| m.get("content"))
            .cloned()
            .unwrap_or(Value::Null);
        for tool_request in extract_assistant_tool_requests(&content) {
            out.push(AssistantToolRequestWithMessage {
                message: message.clone(),
                tool_request,
            });
        }
    }
    out
}
